### PR TITLE
fix: allow exit spans to have child spans with the same type and subtype

### PR DESCRIPTION
**Pull request by kruskall**
Do not silently drop child spans of exit spans. The spec allows for exit spans to have child spans if they have the same type and subtype. One example of this is http trace requests being child spans of an exit span.

Update exit span test to make sure the feature is working as intended.

Closes https://github.com/elastic/apm-agent-go/issues/1319

### CODE REVIEW

File: span.go [hunk 1]
```diff
@@ -65,10 +65,25 @@ func (tx *Transaction) StartSpan(name, spanType string, parent *Span) *Span {
...
 
+	if opts.parent.IsExitSpan() {
+		arr := strings.SplitN(spanType, ".", 3)
+
+		spanType := arr[0]
+		subType := ""
+
+		if len(arr) > 1 {
+			subType = arr[1]
+		}
```
Comment (axw) [inline span.go:80]: There's already code for splitting type/subtype/action. Should we extract that into a function that can be reused here?

https://github.com/elastic/apm-agent-go/blob/31ed25dde1519583f6a6160f47976be428bace87/span.go#L240-L246

**Review by axw — COMMENTED** [review]
Should we be checking the type/subtype in `Span.End`, rather than at creation time? Type/Subtype are mutable.

Also, should we check that the child span is not an exit span? The spec says that the child span must not have any destination context. That's to ensure we don't count both the exit span and its children in destination metrics.

What should happen if a child is created for a child of an exit span?

File: span.go (hunk 1 above)
Comment (kruskall) [reply span.go]: This no longer applies

**Comment by kruskall** [comment]
> Should we be checking the type/subtype in `Span.End`, rather than at creation time? Type/Subtype are mutable.

Good point! Changed.


> The spec says that the child span must not have any destination context. That's to ensure we don't count both the exit span and its children in destination metrics.

Updated the code to remove destination context :+1:  

> Also, should we check that the child span is not an exit span? 
> What should happen if a child is created for a child of an exit span?

The spec does not forbid that. The status of the child span has no effect on the decision of dropping the span.

File: span.go [hunk 2]
```diff
@@ -345,6 +346,14 @@ func (s *Span) End() {
 	if s.Type == "" {
 		s.Type = "custom"
 	}
+	if s.parent.IsExitSpan() {
+		if s.Type != s.parent.Type || s.Subtype != s.parent.Subtype {
+			s.tracer = nil
+			return
+		}
+
+		s.Context.destination = model.DestinationSpanContext{}
```
Comment (axw) [inline span.go:355]:
```suggestion
		s.Context.model.Destination = nil
		s.Context.model.Service = nil
```

File: span.go [hunk 3]
```diff
@@ -345,6 +346,14 @@ func (s *Span) End() {
 	if s.Type == "" {
 		s.Type = "custom"
 	}
+	if s.parent.IsExitSpan() {
+		if s.Type != s.parent.Type || s.Subtype != s.parent.Subtype {
+			s.tracer = nil
```
[the remark this answers is not in the document]
Comment (kruskall) [reply span.go]: Good point! I think we can just follow the normal control flow. I've updated the code

**Review by axw — APPROVED** [review]

**Pull request merged**